Revise data chapter (#856)
hadley/r-pkgs: Data.Rmd

Here are some examples of this type of package data:

-   [tidyr](https://tidyr.tidyverse.org/reference/index.html#data): `billboard` (song rankings), `who` (tuberculosis data from the World Health Organization)
-   [dplyr](https://dplyr.tidyverse.org/reference/index.html#data): `starwars` (Star Wars characters), `storms` (storm tracks)

At the other extreme, some packages exist solely for the purpose of distributing data, along with its documentation.
These are sometimes called "data packages".
A data package can be a nice way to share example data across multiple packages.
It is also a useful technique for getting relatively large, static files out of a more function-oriented package, which might require more frequent updates.
Here are some examples of data packages:

-   [nycflights13](https://nycflights13.tidyverse.org)
-   [babynames](https://github.com/hadley/babynames)

Finally, many packages benefit from having internal data that is used for internal purposes, but that is not directly exposed to the users of the package.

In this chapter we describe useful mechanisms for including data in your package.

> The practical details differ depending on who needs access to the data and what they will do with it:

Reviewer: And how often it changes?

> -   If you want to store binary data (think: an R object) and make it available to the user, put it in `data/`.

Reviewer: ```suggestion
-   If you want to store R objects and make them available to the user, put it in `data/`.
```

    This is the best place to put example datasets.
    All the concrete examples above for data in a package and data as a package use this mechanism.

> -   If you want to store static binary data, but not make it directly available to the user, put it in `R/sysdata.rda`.

Reviewer: ```suggestion
-   If you want to store R objects for your own use as a developer, put it in `R/sysdata.rda`.
```

    This is the best place to put internal data that your functions need.
    <!--EXAMPLES TO USE SOMEWHERE: readr + data-raw/date-symbols.R + date_symbols, googledrive/googlesheets + data-raw/discovery-doc-prep + .endpoints & .schemas & .tidy_schemas-->

-   If you want to store data in some raw, non-R-specific form and make it available to the user, put it in `inst/extdata`.
    For example, readr and readxl each use this mechanism to provide a collection of delimited files and Excel workbooks, respectively.

> -   If you want to store data that reflects your package's internal state, use an environment.

Reviewer: ```suggestion
-   If you want to store data varies over time to reflect your package's internal state, use an environment.
```

Author: I went with

>  If you want to store dynamic data that reflects the internal state of your package within a single R session, use an environment.

    This technique is not as common or well-known as those above, but can be very useful in specific situations.

> -   If you want to store data persistently across R sessions, such as configuration or user-specific data, there are officially sanctioned locations for this.

Reviewer: ```suggestion
-   If you want to store data persistently across R sessions, such as configuration or user-specific data, use one of the officially sanctioned locations.
```

The default is "bzip2", but sometimes "gzip" or "xz" can create smaller files.

It's possible to use other types of files below `data/`, but we don't recommend it because `.rda` files are already fast, small, and explicit.
The other possibilities are described in the documentation for `utils::data()` and in the [Data in packages](https://rstudio.github.io/r-manuals/r-exts/Creating-R-packages.html#data-in-packages) section of Writing R Extensions.
In terms of advice to package authors, the help topic for `data()` seems to implicitly make the same recommendations as we do above:

-   Store one R object in each `data/*.rda` file
-   Use the same name for that object and its `.rda` file
-   Use lazy-loading, by default

If the `DESCRIPTION` contains `LazyData: true`, then datasets will be lazily loaded.
This means that they won't occupy any memory until you use them.
The following example shows memory usage before and after loading the nycflights13 package.
You can see that memory usage doesn't change until you inspect the `flights` dataset stored inside the package.

```{r}

> pryr::mem_used()

Reviewer: ```suggestion
lobstr::mem_used()
```

library(nycflights13)
pryr::mem_used()

invisible(flights)
pryr::mem_used()
```

We recommend that you include `LazyData: true` in your `DESCRIPTION` if you are shipping `.rda` files below `data/`.
If you use `use_data()` to create such datasets, it will automatically make this modification to `DESCRIPTION` for you.

> In terms of advice to package users, note that lazily-loaded datasets do **not** need to be pre-loaded.

Reviewer: Somehow pull this out in a call out that acknowledges the historical important of `data()`, but that you should never use it now.

Author: done

Above, once we did `library(nycflights13)`, we could immediately access `flights`.
There is no call `data(flights)`, because it is not necessary.

There are specific downsides to `data(some_pkg_data)` calls that support a policy of only using `data()` when it is actually necessary, i.e. for datasets that would not be available otherwise:

-   By default `data()`, creates objects in the user's global workspace. There is the potential to silently overwrite pre-existing objects with new values.
-   There is also no guarantee that `data(foo)` will create exactly one object named "foo". It could create more than one object and/or objects with totally different names.

### Preserve the origin story of package data {#sec-data-data-raw}

Often, the data you include in `data/` is a cleaned up version of raw data you've gathered from elsewhere.

> We highly recommend taking the time to include the code used to do this in the source version of your package.

Reviewer: If you want a counter-example: we have no way to recreate the ggplot2 datasets. They're from scripts that existed on Hadley's computer at some time in the past, but not any more.

This makes it easy for you to update or reproduce your version of the data.
This data-creating script is also a natural place to leave comments about important properties of the data, i.e. which features are important for downstream usage in package documentation.

We suggest that you keep this code in one or more `.R` files below `data-raw/`.
You don't want it in the bundled version of your package, so this folder should be listed in `.Rbuildignore`.
usethis has a convenience function that can be called when you first adopt the `data-raw/` practice or when you add an additional `.R` file to the folder:

```{r, eval = FALSE}
usethis::use_data_raw()

usethis::use_data_raw("my_pkg_data")
```

`use_data_raw()` creates the `data-raw/` folder and lists it in `.Rbuildignore`.
A typical script in `data-raw/` includes code to prepare a dataset and ends with a call to `use_data()`.

These data packages all use the approach recommended here for `data-raw/`:

-   [babynames](https://github.com/hadley/babynames)
-   [nycflights13](https://github.com/hadley/nycflights13)
-   [gapminder](https://github.com/jennybc/gapminder)

::: callout-warning
## Submitting to CRAN

> Generally, package data should be smaller than a megabyte - if it's larger you'll need to argue for an exemption.

Reviewer: Call forward to CRAN submission chapter and general challenge of arguing your special case.

In that case, we recommend that you embrace the [UTF-8 Everywhere manifesto](http://utf8everywhere.org) and use the UTF-8 encoding.
The `DESCRIPTION` file placed by `usethis::create_package()` always includes `Encoding: UTF-8`, so by default a devtools-produced package already advertises that it will use UTF-8.

Making sure that the strings embedded in your package data have the intended encoding is something you accomplish in your data preparation code, i.e. in the R scripts below `data-raw/`.
You can use `Encoding()` to learn the current encoding of the elements in a character vector and functions such as `enc2utf8()` or `iconv()` to convert between encodings.

::: callout-warning
## Submitting to CRAN

If you have UTF-8-encoded strings in your package data, you may see this from `R CMD check`:

    -   checking contents of 'data' directory ... OK
    -   checking data for non-ASCII characters ... NOTE
        Note: found 352 marked UTF-8 strings

This `NOTE` is truly informational, i.e. it is a `NOTE` that you do not need to address or attempt to fix.
As long as you actually intend to have UTF-8 strings in your package data, all is well.

> Ironically, this `NOTE` is actually suppressed by `R CMD check --as-cran`, i.e. by including the `--as-cran` flag, despite the fact that this note does appear in the check results once a package is on CRAN.

Reviewer: Even more ironically, we have recently learned that `--as-cran` is not used by CRAN.

Author: I tweaked the callout to state this more clearly.

By default, `devtools::check()` sets the `--as-cran` flag and therefore does not transmit this `NOTE`.
But you can surface it with `check(cran = FALSE, env_vars = c("_R_CHECK_PACKAGE_DATASETS_SUPPRESS_NOTES_" = "false"))`.
:::

<!-- https://github.com/wch/r-source/blob/f6737799b169710006b040f72f9abc5e09180229/src/library/tools/R/QC.R#L4672 -->

If you seem to have non-ASCII strings in your package data and this is unintentional, `tools::checkRdaFiles()` can be helpful for determining which `.rda` file is the culprit.
*NO, WAIT, HOLD ON, THIS IS NOT ACTUALLY TRUE. This is not what the `ASCII` column of the return value actually means. Maybe you really do have to call `tools:::.check_package_datasets()` to figure out which file(s) to blame?*

## Internal data {#data-sysdata}

Sometimes your package functions need access to pre-computed data.

> If you put these objects in `data/`, they'll also be available to package users, which is not appropriate.

Reviewer: I think it's worth saying something like this:

In most cases you should just create these objects in your R code using `data.frame()` or `c()` as appropriate. However, in some cases, you'll have a very large amount of data and it's easier to store in a separate file.

Author: added

Instead, you can save your package's internal data in `R/sysdata.rda`.

Here are some examples of internal package data:

-   Two colour-related packages, [munsell](https://github.com/cwickham/munsell) and [dichromat](https://cran.r-project.org/web/packages/dichromat/index.html), use `R/sysdata.rda` to store large tables of colour data.
-   [googledrive](https://github.com/tidyverse/googledrive) and [googlesheets4](https://github.com/tidyverse/googlesheets4) wrap the Google Drive and Google Sheets APIs, respectively. Both use `R/sysdata.rda` to store data derived from a so-called [Discovery Document](https://developers.google.com/discovery/v1/reference/apis) which "describes the surface of the API, how to access the API and how API requests and responses are structured".

The easiest way to create `R/sysdata.rda` is to use `usethis::use_data(internal = TRUE)`:

```{r, eval = FALSE}
internal_this <- ...
internal_that <- ...

usethis::use_data(internal_this, internal_that, internal = TRUE)
```

> Unlike `data/`, where we advocate for one `.rda` file per exported data object, you store all of your internal data objects together in the single file `R/sysdata.rda`.

Reviewer: ```suggestion
Unlike `data/`, where you use one `.rda` file per exported data object, you store all of your internal data objects together in the single file `R/sysdata.rda`.
```

-   During development, it works for interactive calls, calls from within the loaded package's namespace, and even for calls originating in dependencies.

```{r error = TRUE}
fs::path_package("extdata", package = "idonotexist")

fs::path_package("extdata", "I_do_not_exist.csv", package = "readr")

fs::path_package("extdata", "chickens.csv", package = "readr")
```

### `pkg_example()` path helpers

We like to offer convenience functions that make example files easy to access.
These are just user-friendly wrappers around `system.file()` or `fs::path_package()`, but can have added features, such as the ability to list the example files.
Here's the definition and some usage of `readxl::read_example()`:

```{r, eval = FALSE}
readxl_example <- function(path = NULL) {
  if (is.null(path)) {
    dir(system.file("extdata", package = "readxl"))
  } else {
    system.file("extdata", path, package = "readxl", mustWork = TRUE)
  }
}
```

```{r}
readxl::readxl_example()

readxl::readxl_example("clippy.xlsx")
```

## Internal state

Sometimes there's information that multiple functions from your package need to access that:

> -   Must be determined at runtime, not build time. It might even be dynamic.

Reviewer: Load time, not run time?

This works because environments have reference semantics (whereas more pedestrian R objects, such as atomic vectors, lists, or data frames have value semantics).

[^data-1]: If you don't know much about R environments and what makes them special, a great resource is the [Environments chapter](https://adv-r.hadley.nz/environments.html) of Advanced R.

Consider a package that can store the user's favorite letters or numbers.
You might start out with code like this in a file below `R/`:

```{r eval = FALSE}
favorite_letters <- letters[1:3]

#' Report my favorite letters
#' @export
mfl <- function() {
  favorite_letters
}

#' Change my favorite letters
#' @export
set_mfl <- function(l = letters[24:26]) {
  favorite_letters <<- l
  favorite_letters
}
```

`favorite_letters` is initialized to ("a", "b", "c") when the package is built.
The user can then inspect `favorite_letters` with `mfl()`, at which point they'll probably want to register *their* favorite letters with `set_mfl()`.
Note that we've used the super assignment operator `<<-` in `set_mfl()` in the hope that this will reach up into the package environment and modify the internal data object `favorite_letters`.

> But a call to `set_mfl()` fails like so:

Reviewer: This code works if you run it in a regular R session, but if you put it inside a package and run it you'll see:

Author: added

```{r, eval = FALSE}
set_mfl(c("j", "f", "b"))
#> Error in set_mfl(c("j", "f", "b")): cannot change value of locked binding for
#> 'favorite_letters'
```

Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).

However, if we maintain our internal state within an internal package environment, we can modify objects contained in the environment and even add completely new objects.
Here's an alternative implementation that uses an internal environment named `the`.

```{r, eval = FALSE}
the <- new.env(parent = emptyenv())
the$favorite_letters <- letters[1:3]

#' Report my favorite letters
#' @export
mfl2 <- function() {
  the$favorite_letters
}

#' Change my favorite letters
#' @export
set_mfl2 <- function(l = letters[24:26]) {

>   the$favorite_letters <- l

Reviewer: If you're going to use a set function as an example, I think you should follow our standard practice and invisibly return the previous value.

Author: done

  the$favorite_letters
}
```

Now a user *can* record their favorite letters:

```{r, eval = FALSE}
mfl2()
#> [1] "a" "b" "c"

set_mfl2(c("j", "f", "b"))
#> [1] "j" "f" "b"

mfl2()
#> [1] "j" "f" "b"
```

We often like to use the name "the" for an internal package environment, because it allows you to refer to the objects inside in a very natural way, such as `the$token`, meaning "the token".
It is also important to specify `parent = emptyenv()` for such an internal environment, as you generally don't want the environment to inherit from any other (nonempty) environment.

Here are some examples of how packages use this technique:

-   Various functions in the googledrive package need to know the file ID for the current user's home directory on Google Drive. This is an eye-watering string of 40 seemingly random characters that means nothing to a human. It would be inhumane to expect a user to know this or to pass it into every function. Instead, googledrive determines this ID upon first need, then caches it for later use.
-   TODO: Need 1-2 more examples.

> The blog post [Package-Wide Variables/Cache in R Packages](https://trestletech.com/2013/04/package-wide-variablescache-in-r-package/) gives a more detailed development of this technique.

Reviewer: Worth noting that `load_all()` will wipe out the every time.

I wonder if it's worth comparing `the` to `options()`?

Author: I haven't added this yet, because the section feels like it's already getting very long. But I'll keep it mind as we continue to iterate on this.

Author: Correction: I did add the info about `load_all()` but not the comparison to `options()`.

source("common.R")
status("restructuring")
```

## Introduction

It's often useful to include data in a package.
If the primary purpose of a package is to distribute useful functions, example datasets make it easier to write excellent documentation.
These datasets can be hand-crafted to provide compelling use cases for the functions in the package.
Here are some examples of this type of package data:

-   [tidyr](https://tidyr.tidyverse.org/reference/index.html#data): `billboard` (song rankings), `who` (tuberculosis data from the World Health Organization)
-   [dplyr](https://dplyr.tidyverse.org/reference/index.html#data): `starwars` (Star Wars characters), `storms` (storm tracks)

At the other extreme, some packages exist solely for the purpose of distributing data, along with its documentation.
These are sometimes called "data packages".
A data package can be a nice way to share example data across multiple packages.
It is also a useful technique for getting relatively large, static files out of a more function-oriented package, which might require more frequent updates.
Here are some examples of data packages:

-   [nycflights13](https://nycflights13.tidyverse.org)

> -   [babynames](https://github.com/hadley/babynames)

Reviewer 2: And https://github.com/tidymodels/modeldata if you are looking for something a bit different

Author: AFAICT only one of the datasets (`tate_text`) is clearly produced by the recommended strategy (having a `.R` script below `data-raw/`). So I think I'll leave it off for now.

Finally, many packages benefit from having internal data that is used for internal purposes, but that is not directly exposed to the users of the package.

In this chapter we describe useful mechanisms for including data in your package.
The practical details differ depending on who needs access to the data, how often it changes, and what they will do with it:

> -   If you want to store R objects and make them available to the user, put them in `data/`.

Reviewer: Should probably add section numbers/links here.

Author: done

#>   cannot change value of locked binding for 'favorite_letters'
```

Because `favorite_letters` is a regular character vector, modification requires making a copy and rebinding the name `favorite_letters` to this new value.
And that is what's disallowed: you can't change the binding for objects in the package namespace (well, at least not without trying harder than this).
Defining `favorite_letters` this way only works if you will never need to modify it.

However, if we maintain state within an internal package environment, we **can** modify objects contained in the environment (and even add completely new objects).
Here's an alternative implementation that uses an internal environment named "the".

```{r, eval = FALSE}
the <- new.env(parent = emptyenv())
the$favorite_letters <- letters[1:3]

#' Report my favorite letters
#' @export
mfl2 <- function() {
  the$favorite_letters
}

#' Change my favorite letters
#' @export
set_mfl2 <- function(l = letters[24:26]) {
  old <- the$favorite_letters
  the$favorite_letters <- l
  invisible(old)
}
```

Now a user *can* register their favorite letters:

```{r, eval = FALSE}
mfl2()
#> [1] "a" "b" "c"

set_mfl2(c("j", "f", "b"))

> #> [1] "j" "f" "b"

Reviewer 2: Am I crazy or should this be `[1] "a" "b" "c"`, i.e. the "old" values?

Reviewer 2: And if it returns `invisible(old)` we shouldn't see it at all?

Author: Good catch. This is a copy / paste error. I'm transferring output from a toy package that I've got locally.

Yeah here is the correct chunk here. Will fix.

``` r
mfl2()
#> [1] "a" "b" "c"

set_mfl2(c("j", "f", "b"))

mfl2()
#> [1] "j" "f" "b"
```

Author: Yeah I feel like we have quite of bit of interesting "on load" stuff right now. But that it's also a bit too experimental (and getting into the weeds) to work into the book today.

mfl2()
#> [1] "j" "f" "b"
```

Note that this new value for `the$favorite_letters` persists only for the remainder of the current R session (or until the user calls `set_mfl2()` again).
More precisely, the altered state persists only until the next time the package is loaded (including via `load_all()`).
At load time, the environment `the` is reset to an environment containing exactly one object, named `favorite_letters`, with value ("a", "b", "c").
It's like the movie Groundhog Day.
(We'll discuss more persistent package- and user-specific data in the next section.)

Jim Hester introduced our group to the nifty idea of using "the" as the name of an internal package environment.

> This lets you to refer to the objects inside in a very natural way, such as `the$token`, meaning "the token".

Reviewer 2: ```suggestion
This lets you to refer to the objects inside in a very natural way, such as `the$token`, meaning "_the_ token".
```

In my head I read it with emphasis on _the_ to stress that it is a singleton and it might be worth showing that like this

It is also important to specify `parent = emptyenv()` when defining an internal environment, as you generally don't want the environment to inherit from any other (nonempty) environment.

As seen in the example above, the definition of the environment should happen as a top-level assignment in a file below `R/`.
(In particular, this is a legitimate reason to define a non-function at the top-level of a package; see section @sec-code-when-executed for why this should be rare.) As for where to place this definition, there are two considerations:

-   Define it before you use it.
    If other top-level calls refer to the environment, the definition must come first when the package code is being executed at build time.

>     This is why `R/aaa.R` is a common and safe choice.

Reviewer 2: To get around the need to put it in `R/aaa.R`, in dplyr we use a trick where `the <- new_environment()` can be put anywhere, but when we have top-level calls that refer to it, we wrap them in `rlang::on_load()` to run those expressions in the `.onLoad()` hook (which is after `the` has been created)

I think this idea is experimental enough that your current recommendation is best, but thought I'd mention it

https://cs.github.com/tidyverse/dplyr/blob/5aa0b7c4fcd91f87ce0f76821d5fe6c137b4e8a0/R/dplyr.R#L23

https://cs.github.com/tidyverse/dplyr/blob/5aa0b7c4fcd91f87ce0f76821d5fe6c137b4e8a0/R/conditions.R#L287

Author: Yeah I feel like we have quite of bit of interesting "on load" stuff right now. But that it's also a bit too experimental (and getting into the weeds) to work into the book today.

You need to use the official locations for persistent file storage, because it's the responsible and courteous thing to do and also to comply with CRAN policies.
In particular, you can't just write into the user's home directory.
Here's a relevant excerpt from the CRAN policy at the time of writing:

> Packages should not write in the user's home filespace (including clipboards), nor anywhere else on the file system apart from the R session's temporary directory ....
>
> For R version 4.0 or later (hence a version dependency is required or only conditional use is possible), packages may store user-specific data, configuration and cache files in their respective user directories obtained from `tools::R_user_dir()`, provided that by \[sic\] default sizes are kept as small as possible and the contents are actively managed (including removing outdated material).

In R, there are two tools you can use to derive acceptable locations for user data:

-   The `tools::R_user_dir()` function, available in R \>= 4.0.

> -   The [rappdirs package](https://rappdirs.r-lib.org), which is a port of the Python appdirs module, compatible with R \>= 3.5[^data-3].

Reviewer: I think you can drop rappdirs, or just mention it in a footnote.

Author: Moved to footnote
